pmlc/dialect/pxa/transforms/stencil_generic.cc

```diff
@@ -0,0 +1,197 @@
+    const llvm::SmallVector<unsigned, 3> &tensorIDs) {
+  auto currIdx = boundIdxs->size();
+  if (currIdx == semanticIdxCount) {
+    // This is a legal binding, go find a tiling for it
+    llvm::SmallVector<int64_t, 8> currTileSize(semanticIdxCount);
+    RecursiveTileIndex(TensorAndIndexPermutation(tensorIDs, *boundIdxs),
+                       &currTileSize, 0);
+  } else {
+    for (const auto &blockArg : blockArgs) {
+      // Don't bind same index twice
+      if (std::find(boundIdxs->begin(), boundIdxs->end(), blockArg) !=
+          boundIdxs->end()) {
+        continue;
+      }
+
+      // Verify the requirements for this index with each tensor are all met
+      bool reqsMet = true;
+      for (unsigned i = 0; i < tensorIDs.size(); i++) {
+        auto it = requirements.find(std::make_pair(i, currIdx));
+        if (it != requirements.end() && !it->second(tensorIDs[i], blockArg)) {
+          reqsMet = false;
+          break;
+        }
+      }
+      if (!reqsMet) {
+        continue;
+      }
+
+      // If we made it to here, this index has appropriate semantics; bind it
+      // and recurse
+      boundIdxs->push_back(blockArg);
+      RecursiveBindIndex(boundIdxs, tensorIDs);
+      boundIdxs->pop_back();
+    }
+  }
+}
+
+void StencilGeneric::RecursiveTileIndex(     //
+    const TensorAndIndexPermutation &perm,   //
+    llvm::SmallVector<int64_t, 8> *tileSize, //
+    int64_t currIdx) {
+  assert(tileSize->size() == semanticIdxCount);
+  if (currIdx == semanticIdxCount) {
+    auto cost = getCost(perm, *tileSize);
+    if (VLOG_IS_ON(3)) {
+      std::stringstream currTilingStr;
+      currTilingStr << "[ ";
+      for (const auto &sz : *tileSize) {
+        currTilingStr << sz << " ";
+      }
+      currTilingStr << "]";
+      IVLOG(3, "Considering Tiling " << currTilingStr.str()
+                                     << ", which would have cost " << cost);
+    }
+    if (cost < bestCost) {
+      bestCost = cost;
+      bestPermutation = perm;
+      bestTiling = llvm::SmallVector<int64_t, 8>(tileSize->size());
+      for (size_t i = 0; i < tileSize->size(); i++) {
+        bestTiling[i] = (*tileSize)[i];
+      }
+    }
+  } else {
+    // TODO: Setup cache for the generator
+    for (int64_t currIdxTileSize : tilingGenerators[currIdx](
+             ranges[perm.indexes[currIdx].getArgNumber()])) {
+      (*tileSize)[currIdx] = currIdxTileSize;
+      RecursiveTileIndex(perm, tileSize, currIdx + 1);
+    }
+  }
+}
+
+void StencilGeneric::DoStenciling() {
+  // Initialization
+  auto maybeRanges = op.getConstantRanges();
+  if (maybeRanges) {
+    ranges = maybeRanges.getValue(); // TODO: Is this how to use Optional?
+  } else {
+    IVLOG(4, "Cannot Stencil: Requires constant ranges");
+    return;
+  }
+
+  auto maybeLoadsAndStores = capture();
+  if (maybeLoadsAndStores) {
+    loadsAndStores =
+        maybeLoadsAndStores.getValue(); // TODO: Is this how to use Optional?
```

Comment:
Yes, or there's a `operator*` overload that you can use, or `operator->`, so you can say (just examples or optional usage):

```
loadAndStores = *maybeLoadAndStores;
maybeLoadAndStores->loads.size();
```